### PR TITLE
build: move ioredis as peer dependency

### PR DESCRIPTION
**Pull request by Kikobeats**
Moved `ioredis` as peer dependency to avoid be locked as part of the module.

This give the same functionality plus you can control `ioredis` version out of the module.

### CODE REVIEW

**Comment by lukechilds** [comment]
In theory I don't have a problem with this but in practise it seems a little risky.

Because Keyv is at the database abstraction layer we have to be very conservative about making changes.

This is a pretty vague requirement:

```
`ioredis": "4"
```

What if the user installs a version of `ioredis` that breaks API compat? It could have pretty bad consequences.

Plus all new installs are going to be getting a very up to date version of `ioredis@v4.x.x` which may not have been tested against Keyv.

semver should help here but semver is hard to follow perfectly, breaking changes accidentally creep through.

If this was like a string handling lib or something and the worse case is strings look slightly strange then it's not so much of an issue. At the DB layer it's much more serious.

What would be ideal would be if we could include our own `ioredis` for Keyv and let the application use their own version separately. I don't think npm allows for this though. What happens for you currently, does your app just use the lowest version which is presumable Keyv's?

If we were to merge this we'd also need to update installation instructions on the Keyv repo and we should probably do it for all storage adapters, not just `keyv-redis`.

**Comment by Kikobeats** [comment]
> What if the user installs a version of ioredis that breaks API compat? It could have pretty bad consequences.

Since the `ioredis` is out of the module, you can use any `ioredis@4` that works for you.

> Plus all new installs are going to be getting a very up to date version of ioredis@v4.x.x which may not have been tested against Keyv.

Using `ioredis@v4` you can be sure tests are running using the latest version, so when a new version is released you can quickly detect rather than maintain this repository constantly out of date cause need to be updated dependencies regularly.

Also, it could be complementary with things like a Travis cronjob to be run tests every day is a trivial thing.

> SemVer should help here but SemVer is hard to follow perfectly, breaking changes accidentally creep through.

That happens all the time and it's another point to merge this PR: if you lock a specific `ioredis` dependency and that version has a security problem or whatever, I have to wait you until a new patched `keyv-redis` version will release. However, using this workflow that moves the database dependency of the library I can full control over the version, making the patch just instant.

> Of course we could fork ioredis and use our fork in Keyv so the user has full control over their own ioredis version, but that's a pretty messy solution.

The idea is to reduce the manual job rather than add more job from our side, so I don't think so.

> If we were to merge this we'd also need to update installation instructions on the Keyv repo and we should probably do it for all storage adapters, not just keyv-redis.

Just release this as `keyv-redis@3` about the "breaking change" should enough. After that, it can be replicated to the rest of the adapters.

**Comment by lukechilds** [comment]
>Since the ioredis is out of the module, you can use any ioredis@4 that works for you.

I mean from the Keyv perspective not the user perspective. The user could install  an `ioredis` version that doesn't work with Keyv.

>Using ioredis@v4 you can be sure tests are running using the latest version, so when a new version is released you can quickly detect rather than maintain this repository constantly out of date cause need to be updated dependencies regularly.

This isn't a desirable property for a testing environment though. Tests should be reproducible. Otherwise we don't actually know what version we're testing against until the tests have finished.

e.g If someone is working on a feature branch for a PR, then a new version of `ioredis` gets published, their next commit could fail to build due to `ioredis`, even though there are no errors introduced in the PR branch.

>Also, it could be complementary with things like a Travis cronjob to be run tests every day is a trivial thing.

This is definitely a good idea. If we merge this we should have a separate cron job testing against latest version in the peer dep range.

>if you lock a specific ioredis dependency and that version has a security problem or whatever, I have to wait you until a new patched keyv-redis version will release. However, using this workflow that moves the database dependency of the library I can full control over the version, making the patch just instant.

I completely understand your point here. I understand this is awkward from the consumer end. But in that scenario, it's a pretty simple fix. Just submit a PR to `keyv-redis`, if tests pass I can merge it right away, there's no need to test anything further.

However the alternative is not that simple, if everyone is installing the latest release of `ioredis`, it's actually quite difficult to handle an issue caused by that.

We can detect it with a cron job but that just notifies us there's an issue, it doesn't fix the issue. If it requires special handling inside Keyv to resolve we have to drop what we're doing and fix it asap, before any new users install or current users trigger the bug.

There's also no way for us to alert the users who already installed it that it may behave unpredictably, we just need to hope they update before something bad happens.

>Just release this as keyv-redis@3 and warn users about the "breaking change" should enough. After that, it can be replicated to the rest of the adapters.

I just mean currently on the Keyv repo we have stuff like:


>By default everything is stored in memory, you can optionally also install a storage adapter.
>```
>npm install --save @keyv/redis
>npm install --save @keyv/mongo
>npm install --save @keyv/sqlite
>npm install --save @keyv/postgres
>npm install --save @keyv/mysql
>```

Which would be incorrect as soon as we publish this change so we should update them there first.

---

Sorry for the lengthy response, I hope you understand my concern.

I completely understand your point and I agree it would be great for users to have complete control over the DB client lib they use in their application code. I would like to merge something that enables that.

It's just I'm already struggling to keep up with OSS maintenance as it is and I worry this solution has the potential to introduce some fairly time sensitive bugs that I'm not sure I'd be able to commit to resolving quickly.

**Comment by lukechilds** [comment]
Also, I don't think we need to use a `peerDependencies` to achieve this.

We could just relax the range in `dependencies` to achieve the same result.

**Comment by lukechilds** [comment]
We could probably achieve what you want without any additional risk by doing:

Where `4.16.0` is the hardcoded latest version.

```
"ioredis": ">=4.0.0 <4.16.0 || ~4.16.0"
```

This allows any `ioredis` packages in the `4.x` range up to the latest version. Will automatically get new patches (`4.16.x`) when they're published. But would require us to manually update for SemVer minor updates (`4.x`).

However we could have a cron that runs every 24 hours that checks for a new minor release, and if it exists, run tests against it, and if they pass automatically open a PR to update the dependency range.

I think that's better because:
- No messing around with peer deps, just npm install and you have everything
- npm install always installs the latest version that unit tests have ran against (plus any new patches)
- No added risk over current setup
- Easy to keep up to date once cron job is implemented
- Still gives application authors full control over which package they use in the defined range.

**Comment by WoLfulus** [comment]
The problem with it not being a peer dependency is this line:

https://github.com/lukechilds/keyv-redis/blob/master/src/index.js#L10

Once you pass an existing ioredis instance, the instanceof check fails and we broke our servers because it now fallback to connect to `127.0.0.1` 

:(

**Comment by rijkvanzanten** [comment]
@lukechilds In the current setup, the following example doesn't work:

```js
const KeyvRedis = require('@keyv/redis');
const Redis = require('ioredis');

const myRedis = new Redis();
const cache = new KeyvRedis(myRedis);
```

This is because our installed version of `ioredis@4.19.2` is a different module than `ioredis@4.17.2` as required by `@keyv/redis`. This in turn means that the following check fails:

https://github.com/lukechilds/keyv-redis/blob/9d650f21ce4609c07d5e6848efb1572eccf95d46/src/index.js#L10-L12

As "your" copy of `Redis` is a different object constructor from "ours", that check is always false.

By updating the `ioredis` dependency to a peerDependency with a lock on the `4` major version, this problem is resolved

**Comment by Kikobeats** [comment]
done at https://github.com/microlinkhq/keyv-redis

**Comment by rijkvanzanten** [comment]
@Kikobeats Why did you close this PR? Isn't `@microlink/keyv-redis` a different package than `@keyv/redis`? I'd rather not change to an external fork if there's an "official" solution available..

@lukechilds Can you add one of us as a contributor here / on npm? Would be a shame if these contributions fragment out in published forks

**Comment by Kikobeats** [comment]
@rijkvanzanten that was my original idea, but almost one year later I'm haven't npm publish permissions and the original author not seems too interested in maintaining this up to date.
